### PR TITLE
Remove job_type from bgw_job table

### PR DESCRIPTION
**Pull request by svenklemm**

### CODE REVIEW

File: src/bgw/job_stat.c [hunk 1]
```diff
@@ -238,6 +238,7 @@ calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failure
 	}
 	PG_CATCH();
 	{
+		MemoryContextSwitchTo(oldctx);
```
Comment (mkindahl) [inline src/bgw/job_stat.c:241]: Not sure how this is related to removing job_type. It might be necessary to add this line because of a bug, but I have no way of understanding why this line is necessary since I have no context. Either put it in a separate commit that explains why it was added, or add some sort of comment here (I would prefer the former).

File: sql/updates/latest-dev.sql [hunk 2]
```diff
@@ -24,6 +24,7 @@ DROP FUNCTION IF EXISTS alter_job_schedule;
 
 DROP VIEW IF EXISTS timescaledb_information.policy_stats;
 DROP VIEW IF EXISTS timescaledb_information.drop_chunks_policies;
+DROP VIEW IF EXISTS timescaledb_information.drop_chunks_policies;
```
Comment (mkindahl) [inline sql/updates/latest-dev.sql:27]: Duplicate line?

**Review by mkindahl — CHANGES_REQUESTED** [review]
I'm not sure if this is work in progress or ready to be reviewed. Some immediate things I think you need to look over.

* There is no description with the commit. Please add one.
* There appears to be fixes to bugs intermingled with the code, and I have no idea what those bugs are and can't decide if the changes actually solves the problem since I lack context.
* The `bgw_job_delete` test is removed, but I do not understand why.

File: src/bgw/job.c [hunk 3]
```diff
@@ -603,63 +578,44 @@ ts_bgw_job_validate_job_owner(Oid owner, JobType type)
...
 						rform->rolname.data),
 				 errhint("Hypertable owner must have LOGIN permission to run background tasks.")));
 	}
 	ReleaseSysCache(role_tup);
 }
 
+static bool
+is_telemetry_job(BgwJob *job)
+{
+	return namestrcmp(&job->fd.proc_name, "policy_telemetry_proc") == 0;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/bgw/job.c:591]: This should probably include namespace/schema as well and maybe check that this function is part of our extension.

**Review by erimatnor — CHANGES_REQUESTED** [review]
Most of the changes in this PR look good, although I struggle to understand some of the changes, in particular tests, since I can't see any obvious connection to the removal of `job_type`. I suggest adding more comments to explain what the tests are doing and how it connects to `job_type`. If there's no connection, I suggest splitting this up across several PRs (or at least commits) that each does something distinct.

File: sql/updates/latest-dev.sql [hunk 4]
```diff
@@ -219,6 +214,34 @@ DROP VIEW IF EXISTS timescaledb_information.continuous_aggregates;
 DROP VIEW IF EXISTS timescaledb_information.continuous_aggregate_stats;
 ALTER TABLE IF EXISTS _timescaledb_catalog.continuous_agg DROP COLUMN IF EXISTS job_id;
 
+-- rebuild bgw_job table
+CREATE TABLE _timescaledb_config.bgw_job_tmp AS SELECT * FROM _timescaledb_config.bgw_job;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply sql/updates/latest-dev.sql:218]: renaming a table does not rename objects created with the table (constraints, indexes) so a create table statement while the renamed table exists will not be able to use those names and have different names for those objects.

File: src/bgw/job_stat.c (hunk 1 above)
Comment (svenklemm) [reply src/bgw/job_stat.c:241]: moved into separate PR

File: tsl/src/bgw_policy/job.c [hunk 5]
```diff
@@ -479,7 +479,11 @@ job_execute(BgwJob *job)
 	 */
 	MemoryContextSwitchTo(parent_ctx);
 	arg1 = makeConst(INT4OID, -1, InvalidOid, 4, Int32GetDatum(job->fd.id), false, true);
-	arg2 = makeConst(JSONBOID, -1, InvalidOid, -1, JsonbPGetDatum(job->fd.config), false, false);
+	if (job->fd.config == NULL)
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/bgw_policy/job.c:482]: moved to separate PR

File: sql/views.sql [hunk 6]
```diff
@@ -51,27 +51,12 @@ CREATE OR REPLACE VIEW timescaledb_information.drop_chunks_policies as
...
-SELECT format('%1$I.%2$I', ht.schema_name, ht.table_name)::regclass AS hypertable,
-  jsonb_object_field_text (j.config, 'index_name') AS hypertable_index_name,
-  j.id AS job_id,
-  j.schedule_interval,
-  j.max_runtime,
-  j.max_retries,
-  j.retry_period
-FROM _timescaledb_config.bgw_job j
-  INNER JOIN _timescaledb_catalog.hypertable ht ON j.hypertable_id = ht.id
-WHERE job_type = 'reorder';
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply sql/views.sql:64]: because its a job_type-specific view

File: src/bgw/job.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/bgw/job.c:591]: we use this just as marker for the telemetry job we don't execute the procedure though but call a hardcoded internal telemetry c function so there are no security implications, otherwise people could replace the telemetry procedure and get their own functions executed

File: .travis.yml [hunk 7]
```diff
@@ -45,7 +45,7 @@ jobs:
         - docker exec -u postgres -it docker_arm_emulator find /build -name postmaster.log -exec cat {} +
       script:
       # allow 50 mins to run
-        - travis_wait 50 docker exec -u postgres docker_arm_emulator /bin/bash -c "cd /build && make installcheck IGNORES='multi_transaction_indexing bgw_db_scheduler bgw_job_delete continuous_aggs_insert continuous_aggs_bgw plan_ordered_append-11 parallel-11 compression_ddl continuous_aggs_insert continuous_aggs_multi'"
+        - travis_wait 50 docker exec -u postgres docker_arm_emulator /bin/bash -c "cd /build && make installcheck IGNORES='multi_transaction_indexing bgw_db_scheduler continuous_aggs_insert continuous_aggs_bgw plan_ordered_append-11 parallel-11 compression_ddl continuous_aggs_insert continuous_aggs_multi'"
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply .travis.yml:48]: that test is getting removed, job deletion is moved to bgw_db_scheduler

**Comment by svenklemm** [comment]
> Most of the changes in this PR look good, although I struggle to understand some of the changes, in particular tests, since I can't see any obvious connection to the removal of `job_type`. I suggest adding more comments to explain what the tests are doing and how it connects to `job_type`. If there's no connection, I suggest splitting this up across several PRs (or at least commits) that each does something distinct.

I was trying to get rid of the test hook in the scheduler along with the other changes cause we don't really need it any more but unfortunately that required more changes than i liked so i will do that in separate PR. The job_delete test is moved to the bgw_db_scheduler test as we dont need an isolation test to test it since the execution already is a separate process.

File: tsl/test/expected/continuous_aggs_bgw.out [hunk 8]
```diff
@@ -139,16 +139,17 @@ SELECT ts_bgw_db_scheduler_test_run_and_wait_for_scheduler_finish(25);
...
 
 SELECT * FROM sorted_bgw_log;
- msg_no | mock_time | application_name |                    msg                     
---------+-----------+------------------+--------------------------------------------
-      0 |         0 | DB Scheduler     | [TESTING] Registered new background worker
-      1 |         0 | DB Scheduler     | [TESTING] Wait until 25000, started at 0
-(2 rows)
+ msg_no | mock_time |          application_name          |                                                     msg                                                      
+--------+-----------+------------------------------------+--------------------------------------------------------------------------------------------------------------
+      0 |         0 | Continuous Aggregate Policy [1000] | materializing continuous aggregate public.test_continuous_agg_view: nothing to invalidate, new range up to 6
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_bgw.out:144]: Why do we see this extra job here now? I guess I don't expect the behavior to change when job_type was removed.

File: tsl/test/expected/continuous_aggs_bgw.out [hunk 9]
```diff
@@ -367,11 +368,13 @@ SELECT ts_bgw_db_scheduler_test_run_and_wait_for_scheduler_finish(25);
...
 
 SELECT * FROM sorted_bgw_log;
- msg_no | mock_time | application_name |                    msg                     
---------+-----------+------------------+--------------------------------------------
-      0 |         0 | DB Scheduler     | [TESTING] Registered new background worker
-      1 |         0 | DB Scheduler     | [TESTING] Wait until 25000, started at 0
-(2 rows)
+ msg_no | mock_time |          application_name          |                                                                 msg                                                                  
+--------+-----------+------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------
+      0 |         0 | Continuous Aggregate Policy [1001] | new materialization range for public.test_continuous_agg_table (time column time) larger than allowed in one run, truncating 12 to 2
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_bgw.out:373]: Output also different here. Previously the caggs policy didn't seem to exist here, now it does. Why? Was there previously some filtering on job_type that no longer happens?

File: tsl/test/expected/tsl_tables.out [hunk 10]
```diff
@@ -158,23 +154,25 @@ WARNING:  could not add retention policy due to existing policy on hypertable wi
...
- 1002 | Retention Policy [1002] | drop_chunks | @ 1 day           | @ 5 mins    |          -1 | @ 5 mins     | policy_retention | _timescaledb_internal | default_perm_user | t         |             1 | {"hypertable_id": 1, "retention_window": "@ 3 mons"}
-(1 row)
+SELECT * FROM _timescaledb_config.bgw_job WHERE id >= 1000 ORDER BY id;
+  id  |    application_name     | schedule_interval | max_runtime | max_retries | retry_period |    proc_name     |      proc_schema      |       owner       | scheduled | hypertable_id |                          config                           
+------+-------------------------+-------------------+-------------+-------------+--------------+------------------+-----------------------+-------------------+-----------+---------------+-----------------------------------------------------------
+ 1000 | Reorder Policy [1000]   | @ 84 hours        | @ 0         |          -1 | @ 5 mins     | policy_reorder   | _timescaledb_internal | default_perm_user | t         |             1 | {"index_name": "test_table_time_idx", "hypertable_id": 1}
```
Comment (erimatnor) [inline tsl/test/expected/tsl_tables.out:160]: This job wasn't here before, so behavior seems to have changed. Is this correct/expected? Could you explain?

**Review by erimatnor — COMMENTED** [review]
While the changes related to the removal of job_type seems relative straightforward, I still don't understand the large changes in test behavior.

There are many tests that now display more jobs and log entries from jobs that previously, which indicates that the behavior changed significantly. I don't follow how this happened as a result of removing job_type. There's no PR description attach to explain the changes or the new behavior so as a reviewer I am left to guessing or trying to figure things out myself, which is quite difficult.

Can you please provide more context and a description of the changes and why the test behavior has changed so much.

File: tsl/test/expected/continuous_aggs_bgw.out (hunk 8 above)
Comment (svenklemm) [reply tsl/test/expected/continuous_aggs_bgw.out:144]: this is because policy jobs wouldnt get executed in the scheduler test environment so their output was insert into bgw_log

File: tsl/test/expected/continuous_aggs_bgw.out (hunk 9 above)
Comment (svenklemm) [reply tsl/test/expected/continuous_aggs_bgw.out:373]: same as above policy output wasnt redirected to bgw_log before but now is because we dont differentiate between test jobs and normal jobs anymore

File: tsl/test/expected/continuous_aggs_bgw.out (hunk 8 above)
Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_bgw.out:144]: are you going to filter these again, like discussed?

Comment (svenklemm) [reply tsl/test/expected/continuous_aggs_bgw.out:144]: i added filter for those queries that selected from bgw_job with a job_type filter, but for bgw_log i think it's correct to include log output from the job run

File: src/bgw/job.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/bgw/job.c]: Ok, so then there is at least not a security issue, but this is still quite obscure. Could there be some other way to indicate that this is the telemetry job that is less obscure? Set this column to NULL and use the `application_name` column?

File: test/src/bgw/scheduler_mock.c [hunk 11]
```diff
@@ -272,56 +287,47 @@ test_job_3_long()
...
-	StartTransactionCommand();
-	elog(WARNING, "Execute job 4");
-	CommitTransactionCommand();
-	return true;
+	PG_RETURN_VOID();
 }
 
 static bool
-test_job_5_lock()
+test_job_3_long()
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply test/src/bgw/scheduler_mock.c:295]: this test was not added by me

**Review by erimatnor — APPROVED** [review]

File: sql/updates/latest-dev.sql [hunk 12]
```diff
@@ -11,7 +11,7 @@ DROP FUNCTION _timescaledb_internal.partitioning_column_to_pretty;
 DROP FUNCTION _timescaledb_internal.range_value_to_pretty;
 -- end of do not reorder
 DROP VIEW IF EXISTS timescaledb_information.compressed_chunk_stats;
-DROP VIEW If EXISTS timescaledb_information.compressed_hypertable_stats;
+DROP VIEW IF EXISTS timescaledb_information.compressed_hypertable_stats;
```
Comment (pmwkaa) [inline sql/updates/latest-dev.sql:14]: Quick question: how this update script is tested now?

**Review by pmwkaa — APPROVED** [review]
Looks much cleaner without job_type, could not find any issues right away

File: sql/updates/latest-dev.sql (hunk 12 above)
Comment (svenklemm) [reply sql/updates/latest-dev.sql:14]: it is run during update test

File: src/bgw/job.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/bgw/job.c]: well I disagree proc_name/proc_schema determine what gets executed just that we dont do lookup here but go straight to a c function

File: src/bgw/job.c [hunk 13]
```diff
@@ -603,63 +578,43 @@ ts_bgw_job_validate_job_owner(Oid owner, JobType type)
...
 	ReleaseSysCache(role_tup);
 }
 
+static bool
+is_telemetry_job(BgwJob *job)
+{
+	return namestrcmp(&job->fd.proc_schema, INTERNAL_SCHEMA_NAME) == 0 &&
+		   namestrcmp(&job->fd.proc_name, "policy_telemetry") == 0;
+}
+
```
Comment (mkindahl) [inline src/bgw/job.c:594]: Still find this an odd way to designate the telemetry job. Don't have a better suggestion though.

**Review by mkindahl — APPROVED** [review]
